// privategpt-sdk-utils/Client.js
/**
 * This file was auto-generated by Fern from our API Definition.
 */
import { ContextChunks } from './api/resources/contextChunks/client/Client';
import { ContextualCompletions } from './api/resources/contextualCompletions/client/Client';
import { Embeddings } from './api/resources/embeddings/client/Client';
import { Health } from './api/resources/health/client/Client';
import { Ingestion } from './api/resources/ingestion/client/Client';
import { Recipes } from './api/resources/recipes/Client';
export class PrivategptApiClient {
    _options;
    constructor(_options) {
        this._options = _options;
    }
    _contextualCompletions;
    get contextualCompletions() {
        return (this._contextualCompletions ??= new ContextualCompletions(this._options));
    }
    _contextChunks;
    get contextChunks() {
        return (this._contextChunks ??= new ContextChunks(this._options));
    }
    _ingestion;
    get ingestion() {
        return (this._ingestion ??= new Ingestion(this._options));
    }
    _embeddings;
    get embeddings() {
        return (this._embeddings ??= new Embeddings(this._options));
    }
    _recipes;
    get recipes() {
        return (this._recipes ??= new Recipes(this._options));
    }
    _health;
    get health() {
        return (this._health ??= new Health(this._options));
    }
}
